### PR TITLE
Antenna are supply beacons too. Also refactor supply beacon code

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Antenna helmet modules can place supply beacons for 100 seconds.
Refactor the code for supply beacon a bit
Cooldown for sending supply from 2 minutes to 1

## Why It's Good For The Game

More versatility for RO, will make forward bases more easy to do. Also make antenna a legit module for helmet

## Changelog
:cl:
add: Antenna helmet module can place supply beacons for 100 seconds
balance: Supply drop cooldown from 2 minutes to 1
refactor: Supply drop code is redone a bit
qol: No longer multiple supply pads, only one is needed
/:cl:

### CODE REVIEW

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 1]
```diff
@@ -23,85 +27,58 @@
...
-/obj/item/squad_beacon/bomb/attack_self(mob/living/carbon/human/H)
+/obj/item/squad_beacon/proc/deactivate(mob/living/carbon/human/H)
+	activated = FALSE
+	anchored = FALSE
+	w_class = initial(w_class)
+	layer = initial(layer)
+	name = initial(name)
+	icon_state = initial(icon_state)
+	playsound(src, 'sound/machines/twobeep.ogg', 15, 1)
+	usr.visible_message("[usr] deactivates [src]",
```
Comment (XSlayer300) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:68]:
```suggestion
	usr.visible_message("[usr] deactivates [src].",
```

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 2]
```diff
@@ -23,85 +27,58 @@
...
+	"You deactivate [src]")
```
Comment (XSlayer300) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:69]:
```suggestion
	"You deactivate [src].")
```

File: code/game/objects/machinery/squad_supply/supply_console.dm [hunk 3]
```diff
@@ -6,55 +6,64 @@
...
 	current_beacon = null
 	return ..()
 
 /obj/machinery/computer/supplydrop_console/ui_interact(mob/user, datum/tgui/ui)
 	. = ..()
 	ui = SStgui.try_update_ui(user, src, ui)
+	if(!current_beacon?.registered)
+		set_current_beacon()
+		if(!current_beacon)
+			to_chat(user, "<span class='warning'>No active beacon was detected! </span>")
```
Comment (XSlayer300) [inline code/game/objects/machinery/squad_supply/supply_console.dm:40]:
```suggestion
			to_chat(user, "<span class='warning'>No active beacon was detected!</span>")
```

File: code/game/objects/machinery/squad_supply/supply_console.dm [hunk 4]
```diff
@@ -67,12 +76,11 @@
 		return
 
 	switch(action)
-		if("select_squad")
-			var/datum/squad/selected_squad = SSjob.squads[params["select_squad"]]
-			if(!istype(selected_squad))
+		if("select_beacon")
+			var/selected_beacon = tgui_input_list(ui.user, "Chose which beacon you want to send the supplies", "Beacon choice", GLOB.supply_beacon_list)
```
Comment (XSlayer300) [inline code/game/objects/machinery/squad_supply/supply_console.dm:80]:
```suggestion
			var/selected_beacon = tgui_input_list(ui.user, "Choose which beacon you want to send the supplies.", "Beacon Choice", GLOB.supply_beacon_list)
```

File: code/game/objects/machinery/squad_supply/supply_console.dm [hunk 5]
```diff
@@ -93,56 +101,52 @@
 			if(!COOLDOWN_CHECK(src, next_fire))
 				return
 
-			if(!current_squad?.sbeacon)
-				to_chat(usr, "[icon2html(src, usr)] <span class='warning'>There was an issue with that squads beacon. Check it's still active.</span>")
+			if(!current_beacon)
+				to_chat(usr, "[icon2html(src, usr)] <span class='warning'>There was an issue with that beacon. Check it's still active.</span>")
```
Comment (XSlayer300) [inline code/game/objects/machinery/squad_supply/supply_console.dm:105]:
```suggestion
				to_chat(usr, "[icon2html(src, usr)] <span class='warning'>There was an issue with that beacon. Check if it's still active.</span>")
```

File: code/modules/modular_armor/helmets.dm [hunk 6]
```diff
@@ -111,7 +111,26 @@
...
+
 
 /obj/item/helmet_module/antenna/toggle_module(mob/living/user, obj/item/clothing/head/modular/parent)
 	var/turf/location = get_turf(src)
 	user.show_message("<span class='warning'>The [src] beeps and states, \"Current location coordinates: LONGITUDE [location.x]. LATITUDE [location.y]. Area ID: [get_area(src)]\"</span>", EMOTE_AUDIBLE, "<span class='notice'>The [src] vibrates but you can not hear it!</span>")
+	playsound(user, 'sound/machines/twobeep.ogg', 15, 1)
+	if(beacon.registered)
+		qdel(visual_effect)
+		visual_effect = null
+		user.show_message("<span class='warning'>You remove your last position from the list of possible supply drop locations")
```
Comment (XSlayer300) [inline code/modules/modular_armor/helmets.dm:127]:
```suggestion
		user.show_message("<span class='warning'>You remove your last position from the list of possible supply drop locations.")
```

File: code/modules/modular_armor/helmets.dm [hunk 7]
```diff
@@ -111,7 +111,26 @@
...
+		beacon.unregister()
+		return
+	visual_effect = new(location)
+	beacon.register(user.name + " mobile beacon", location, 100 SECONDS)
+	user.show_message("<span class='warning'>The ship received your transmission, this position will now be available for a supply drop for 100 seconds")
```
Comment (XSlayer300) [inline code/modules/modular_armor/helmets.dm:132]:
```suggestion
	user.show_message("<span class='warning'>The ship received your transmission, this position will now be available for a supply drop for 100 seconds.")
```

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 8]
```diff
@@ -124,7 +124,7 @@
 	var/image/holder = hud_list[SQUAD_HUD]
 	if(!holder)
 		return
-	holder.icon = 'icons/Marine/marine-items.dmi'
+	holder.icon = icon
```
Comment (SplinterGP) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm:127]: whats that for

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 9]
```diff
@@ -9,12 +9,16 @@
 	var/datum/squad/squad = null
 	var/icon_activated = "motion2"
 	var/obj/machinery/camera/beacon_cam = null
+	var/datum/supply_beacon/beacon
+
+/obj/item/squad_beacon/Initialize()
+	. = ..()
+	beacon = new /datum/supply_beacon
```
Comment (SplinterGP) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:16]:
```suggestion
	beacon = new 
```

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 10]
```diff
@@ -23,85 +27,58 @@
...
-		to_chat(user, "<span class='warning'>Your squad already has a beacon activated.</span>")
+	if(!is_ground_level(H.z))
+		to_chat(H, "<span class='warning'>You have to be on the planet to use this or it won't transmit.</span>")
 		return
-	var/area/A = get_area(user)
-	if(A && istype(A) && A.ceiling >= CEILING_METAL)
-		to_chat(user, "<span class='warning'>You have to be outside or under a glass ceiling to activate this.</span>")
+	var/area/A = get_area(H)
+	if(A && istype(A) && A.ceiling >= CEILING_DEEP_UNDERGROUND)
+		to_chat(H, "<span class='warning'>This won't work if you're standing deep underground.</span>")
```
Comment (SplinterGP) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:39]: balance change

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 11]
```diff
@@ -133,7 +118,7 @@
 		cam_name += H.name
 		if(H.assigned_squad)
 			squad = H.assigned_squad
-			name += " ([squad.name])"
+			name += "[squad.name] + squad"
```
Comment (SplinterGP) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:121]: uh?

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 12]
```diff
@@ -173,3 +158,23 @@
...
+	GLOB.supply_beacon_list |= src
+	registered = TRUE
+	if(!life_time)
+		return
+	addtimer(CALLBACK(src, .proc/unregister), life_time)
+
+/datum/supply_beacon/proc/unregister()
+	GLOB.supply_beacon_list -= src
+	SEND_SIGNAL(src, COMSIG_BEACON_UNREGISTERED)
+	registered = FALSE
```
Comment (SplinterGP) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:180]: why is this needed in here and a datum ? make a component

File: code/game/objects/machinery/squad_supply/supply_console.dm [hunk 13]
```diff
@@ -6,55 +6,64 @@
...
 
 /obj/machinery/computer/supplydrop_console/ui_interact(mob/user, datum/tgui/ui)
 	. = ..()
 	ui = SStgui.try_update_ui(user, src, ui)
+	if(!current_beacon?.registered)
+		set_current_beacon()
+		if(!current_beacon)
+			to_chat(user, "<span class='warning'>No active beacon was detected!</span>")
+			ui?.close()
+			return
```
Comment (SplinterGP) [inline code/game/objects/machinery/squad_supply/supply_console.dm:42]: put this before update if done this way.

File: code/game/objects/machinery/squad_supply/supply_console.dm [hunk 14]
```diff
@@ -6,55 +6,64 @@
...
-	.["current_beacon"] = list(
-		"name" = current_squad?.sbeacon?.name,
-		"x_coords" = current_squad?.sbeacon?.loc.x,
-		"y_coords" = current_squad?.sbeacon?.loc.y
-	)
+	.["current_beacon"] = current_beacon
+	.["beacon_x"] = current_beacon?.location.x
+	.["beacon_y"] = current_beacon?.location.y
+	.["beacon_name"] = current_beacon?.name 
+	.["beacon_registered"] = current_beacon?.registered
```
Comment (SplinterGP) [inline code/game/objects/machinery/squad_supply/supply_console.dm:66]: these vars should be static data, only updated when beacon changes.

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/game/objects/machinery/squad_supply/supply_beacon.dm (hunk 12 above)
Comment (BraveMole) [reply code/game/objects/machinery/squad_supply/supply_beacon.dm:180]: why would this need to be a component?

Comment (SplinterGP) [reply code/game/objects/machinery/squad_supply/supply_beacon.dm:180]: because you already made it like one ?

File: code/_globalvars/lists/objects.dm [hunk 15]
```diff
@@ -11,11 +11,12 @@ GLOBAL_LIST_EMPTY(tank_list)
 GLOBAL_LIST_EMPTY(implant_list)
 GLOBAL_LIST_EMPTY(head_list)
 GLOBAL_LIST_EMPTY(beacon_list)
+GLOBAL_LIST_EMPTY(supply_beacon_list)
 GLOBAL_LIST_EMPTY(id_card_list)
 GLOBAL_LIST_EMPTY(disposal_list)
 GLOBAL_LIST_EMPTY(ladder_list)
 GLOBAL_LIST_EMPTY(brig_closets)
-GLOBAL_LIST_EMPTY(supply_pad_list)
+GLOBAL_VAR_INIT(supply_pad, null)
```
Comment (TiviPlus) [inline code/_globalvars/lists/objects.dm:19]: in code/_globalvars/lists/objects.dm ?

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 16]
```diff
@@ -153,3 +153,10 @@
 	icon_state = "bhole3"
 	layer = ABOVE_LYING_MOB_LAYER
 	duration = WRAITH_BANISH_BASE_DURATION+1 //So we don't delete our contents early
+
+/obj/effect/temp_visual/order/temporary_supply_beacon
+	name = "temp supply beacon"
+	icon = 'icons/misc/mark.dmi'
+	icon_state_on = "x3"
```
Comment (TiviPlus) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm:160]: isnt this the AI debug node icon

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 17]
```diff
@@ -153,3 +153,10 @@
...
+	layer = ABOVE_LYING_MOB_LAYER
```
Comment (TiviPlus) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm:161]: why this?

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 18]
```diff
@@ -23,85 +27,58 @@
 		beacon_cam = null
 	return ..()
 
-/obj/item/squad_beacon/attack_self(mob/user)
+/obj/item/squad_beacon/proc/activate(mob/living/carbon/human/H)
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:30]: Autodoc

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 19]
```diff
@@ -23,85 +27,58 @@
...
-/obj/item/squad_beacon/bomb
-	name = "orbital beacon"
-	desc = "A bulky device that fires a beam up to an orbiting vessel to send local coordinates."
-	icon_state = "motion4"
-	w_class = WEIGHT_CLASS_SMALL
-	activation_time = 80
-	icon_activated = "motion1"
-
-/obj/item/squad_beacon/bomb/attack_self(mob/living/carbon/human/H)
+/obj/item/squad_beacon/proc/deactivate(mob/living/carbon/human/H)
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:60]: Autodoc

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 20]
```diff
@@ -23,85 +27,58 @@
...
+	if(do_after(H, delay, TRUE, src, BUSY_ICON_GENERIC))
+		H.transferItemToLoc(src, H.loc)
+		beacon.register(beacon_name, loc)
 		activated = 1
 		anchored = TRUE
 		w_class = 10
 		icon_state = "[icon_activated]"
 		playsound(src, 'sound/machines/twobeep.ogg', 15, 1)
-		user.visible_message("[user] activates [src]",
+		H.visible_message("[H] activates [src]",
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:57]: Nonblocking, but no spans here

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 21]
```diff
@@ -23,85 +27,58 @@
...
-		squad.sbeacon = src
-		user.transferItemToLoc(src, user.loc)
+	var/beacon_name
+	if(H.assigned_squad)
+		beacon_name = H.assigned_squad.name + " beacon"
+	else
+		beacon_name = "Unaffiliated beacon"
+	if(do_after(H, delay, TRUE, src, BUSY_ICON_GENERIC))
+		H.transferItemToLoc(src, H.loc)
+		beacon.register(beacon_name, loc)
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:51]: No early return?

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 22]
```diff
@@ -23,85 +27,58 @@
...
+	anchored = FALSE
+	w_class = initial(w_class)
+	layer = initial(layer)
+	name = initial(name)
+	icon_state = initial(icon_state)
+	playsound(src, 'sound/machines/twobeep.ogg', 15, 1)
+	usr.visible_message("[usr] deactivates [src].",
+	"You deactivate [src].")
+	usr.put_in_active_hand(src)
+	beacon.unregister()
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:71]: usr

File: code/game/objects/machinery/squad_supply/supply_beacon.dm (hunk 11 above)
Comment (TiviPlus) [reply code/game/objects/machinery/squad_supply/supply_beacon.dm:121]: very uh

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 23]
```diff
@@ -173,3 +158,23 @@
 		"You deactivate [src]")
 		H.put_in_active_hand(src)
 
+/datum/supply_beacon
+	var/name = ""
+	var/turf/location
+	var/registered = FALSE
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:164]: Autodoc

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 24]
```diff
@@ -173,3 +158,23 @@
...
+
+/datum/supply_beacon/proc/register(_name, turf/_location, life_time = 0 SECONDS)
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:166]: Autodoc

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 25]
```diff
@@ -173,3 +158,23 @@
...
+		return
+	name = _name
+	location = _location
+	GLOB.supply_beacon_list |= src
+	registered = TRUE
+	if(!life_time)
+		return
+	addtimer(CALLBACK(src, .proc/unregister), life_time)
+
+/datum/supply_beacon/proc/unregister()
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:177]: Autodoc

File: code/game/objects/machinery/squad_supply/supply_console.dm [hunk 26]
```diff
@@ -6,55 +6,64 @@
...
 		ui = new(user, src, "SupplyDropConsole", name)
 		ui.open()
 
-
-/obj/machinery/computer/supplydrop_console/ui_static_data(mob/user)
-	. = ..()
-	.["squads"] = list()
-	for(var/squad in SSjob.active_squads)
-		.["squads"] += list(squad)
+/obj/machinery/computer/supplydrop_console/proc/set_current_beacon(datum/source, datum/supply_beacon/beacon)
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_console.dm:47]: Autodoc

File: code/game/objects/machinery/squad_supply/supply_console.dm (hunk 14 above)
Comment (TiviPlus) [reply code/game/objects/machinery/squad_supply/supply_console.dm:66]: See splinters review

File: code/modules/modular_armor/helmets.dm [hunk 27]
```diff
@@ -111,7 +111,26 @@
 	icon_state = "antenna_head_obj"
 	item_state = "antenna_head"
 	module_type = ARMOR_MODULE_TOGGLE
+	///The supply beacon attached to the module
+	var/datum/supply_beacon/beacon = new
```
Comment (TiviPlus) [inline code/modules/modular_armor/helmets.dm:115]: Do it in Initialize so we dont call init

File: code/modules/modular_armor/helmets.dm [hunk 28]
```diff
@@ -111,7 +111,26 @@
...
+		user.show_message("<span class='warning'>You remove your last position from the list of possible supply drop locations.")
+		beacon.unregister()
+		return
+	visual_effect = new(location)
+	beacon.register(user.name + " mobile beacon", location, 100 SECONDS)
+	user.show_message("<span class='warning'>The ship received your transmission, this position will now be available for a supply drop for 100 seconds.")
+
+/obj/item/helmet_module/antenna/Destroy()
+	. = ..()
+	beacon.unregister()
```
Comment (TiviPlus) [inline code/modules/modular_armor/helmets.dm:136]: Should null it here so it gc's

File: code/modules/modular_armor/helmets.dm [hunk 29]
```diff
@@ -111,7 +111,26 @@
...
+	var/obj/effect/temp_visual/order/temporary_supply_beacon/visual_effect
+
 
 /obj/item/helmet_module/antenna/toggle_module(mob/living/user, obj/item/clothing/head/modular/parent)
 	var/turf/location = get_turf(src)
 	user.show_message("<span class='warning'>The [src] beeps and states, \"Current location coordinates: LONGITUDE [location.x]. LATITUDE [location.y]. Area ID: [get_area(src)]\"</span>", EMOTE_AUDIBLE, "<span class='notice'>The [src] vibrates but you can not hear it!</span>")
+	playsound(user, 'sound/machines/twobeep.ogg', 15, 1)
+	if(beacon.registered)
+		qdel(visual_effect)
+		visual_effect = null
```
Comment (TiviPlus) [inline code/modules/modular_armor/helmets.dm:126]: QDEL_NULL

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/game/objects/machinery/squad_supply/supply_beacon.dm (hunk 12 above)
Comment (TiviPlus) [reply code/game/objects/machinery/squad_supply/supply_beacon.dm:180]: It would use to much getcomponent to be a proper component

File: code/_globalvars/lists/objects.dm (hunk 15 above)
Comment (BraveMole) [reply code/_globalvars/lists/objects.dm:19]: no one will notice

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm (hunk 16 above)
Comment (BraveMole) [reply code/game/objects/effects/temporary_visuals/miscellaneous.dm:160]: no idea what it is used on, i've never seen it in game and it fits pretty well

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm (hunk 17 above)
Comment (BraveMole) [reply code/game/objects/effects/temporary_visuals/miscellaneous.dm:161]: the user of the antenna beacon might be confused to think that the supply is always on him, while no, when he uses the antenna it mark its current location. having a marker on the zone make the feature more intuitive

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 30]
```diff
@@ -159,6 +159,13 @@
 	layer = ABOVE_LYING_MOB_LAYER
 	duration = WRAITH_BANISH_BASE_DURATION+1 //So we don't delete our contents early
 
+/obj/effect/temp_visual/order/temporary_supply_beacon
+	name = "temp supply beacon"
+	icon = 'icons/misc/mark.dmi'
+	icon_state_on = "x3"
+	layer = UNDERFLOOR_OBJ_LAYER
```
Comment (BraveMole) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm:166]: The layer doesn't work when it's added to an hud it seems, so this is pretty ugly

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm [hunk 31]
```diff
@@ -159,6 +159,13 @@
 	layer = ABOVE_LYING_MOB_LAYER
 	duration = WRAITH_BANISH_BASE_DURATION+1 //So we don't delete our contents early
 
+/obj/effect/temp_visual/order/temporary_supply_beacon
+	name = "temp supply beacon"
+	icon = 'icons/misc/mark.dmi'
+	icon_state_on = "x3"
```
Comment (TiviPlus) [inline code/game/objects/effects/temporary_visuals/miscellaneous.dm:165]: Why a pure wight x thats pretty ugly, this was originally Xeno AI debug

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 32]
```diff
@@ -9,12 +9,16 @@
...
 	var/obj/machinery/camera/beacon_cam = null
+	var/datum/supply_beacon/beacon
+
+/obj/item/squad_beacon/Initialize()
+	. = ..()
+	AddComponent(/datum/component/supply_beacon)
 
 /obj/item/squad_beacon/Destroy()
 	GLOB.active_orbital_beacons -= src
+	SEND_SIGNAL(src, COMSIG_UNREGISTER_BEACON)
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:20]: This is not how signals should be used

File: code/modules/modular_armor/helmets.dm [hunk 33]
```diff
@@ -111,7 +111,32 @@
...
+	AddComponent(/datum/component/supply_beacon, CALLBACK(src, .proc/beacon_unregistered))
 
 /obj/item/helmet_module/antenna/toggle_module(mob/living/user, obj/item/clothing/head/modular/parent)
 	var/turf/location = get_turf(src)
 	user.show_message("<span class='warning'>The [src] beeps and states, \"Current location coordinates: LONGITUDE [location.x]. LATITUDE [location.y]. Area ID: [get_area(src)]\"</span>", EMOTE_AUDIBLE, "<span class='notice'>The [src] vibrates but you can not hear it!</span>")
+	playsound(user, 'sound/machines/twobeep.ogg', 15, 1)
+	if(beacon_registered)
+		qdel(visual_effect)
+		user.show_message("<span class='warning'>You remove your last position from the list of possible supply drop locations.")
+		SEND_SIGNAL(src, COMSIG_UNREGISTER_BEACON)
```
Comment (TiviPlus) [inline code/modules/modular_armor/helmets.dm:130]: Ditto

File: code/modules/modular_armor/helmets.dm [hunk 34]
```diff
@@ -111,7 +111,32 @@
...
+	SEND_SIGNAL(src, COMSIG_REGISTER_BEACON ,user.name + " mobile beacon", location, 100 SECONDS)
+	beacon_registered = TRUE
+	user.show_message("<span class='warning'>The ship received your transmission, this position will now be available for a supply drop for 100 seconds.")
+
+/obj/item/helmet_module/antenna/proc/beacon_unregistered()
+	beacon_registered = FALSE
+
+/obj/item/helmet_module/antenna/Destroy()
+	. = ..()
+	SEND_SIGNAL(src, COMSIG_UNREGISTER_BEACON)
```
Comment (TiviPlus) [inline code/modules/modular_armor/helmets.dm:142]: Ditto

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/game/objects/effects/temporary_visuals/miscellaneous.dm (hunk 30 above)
Comment (SplinterGP) [reply code/game/objects/effects/temporary_visuals/miscellaneous.dm:166]: ask slotbot for help he messed with layers

Comment (TiviPlus) [reply code/game/objects/effects/temporary_visuals/miscellaneous.dm:166]: use planes

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 35]
```diff
@@ -23,85 +26,61 @@
...
-	if(squad.sbeacon)
-		to_chat(user, "<span class='warning'>Your squad already has a beacon activated.</span>")
+	if(!is_ground_level(H.z))
+		to_chat(H, "<span class='warning'>You have to be on the planet to use this or it won't transmit.</span>")
 		return
-	var/area/A = get_area(user)
-	if(A && istype(A) && A.ceiling >= CEILING_METAL)
-		to_chat(user, "<span class='warning'>You have to be outside or under a glass ceiling to activate this.</span>")
+	var/area/A = get_area(H)
+	if(A && istype(A) && A.ceiling >= CEILING_DEEP_UNDERGROUND)
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:38]:
```suggestion
	if(isarea(A) && A.ceiling >= CEILING_DEEP_UNDERGROUND)
```

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 36]
```diff
@@ -149,27 +136,76 @@
...
+	unregistered_callback = _unregistered_callback
+	
+///Register this supply beacon into the supply beacon list
+/datum/component/supply_beacon/proc/register(datum/source, _name, turf/_location, life_time = 0 SECONDS)
+	SIGNAL_HANDLER
+	if(registered)
+		return
+	for(var/datum/component/supply_beacon/registered_beacon AS in GLOB.supply_beacon_list)
+		if(registered_beacon.name == _name)
+			return
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:188]: Why des this check exist

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 37]
```diff
@@ -149,27 +136,76 @@
...
+	name = _name
+	location = _location
+	GLOB.supply_beacon_list |= src
+	registered = TRUE
+	RegisterSignal(parent, COMSIG_UNREGISTER_BEACON, .proc/unregister)
+	RegisterSignal(parent, COMSIG_PARENT_QDELETING, .proc/parent_qdeleting)
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:194]: Never unregistered

File: code/game/objects/machinery/squad_supply/supply_console.dm [hunk 38]
```diff
@@ -6,59 +6,69 @@
...
-	var/launch_cooldown = 2 MINUTES
+	var/launch_cooldown = 1 MINUTES
 	COOLDOWN_DECLARE(next_fire)
-	///If true, the last squad will be memorised
-	var/squad_lock = null
-	///The squad wich we will look for beacons
-	var/datum/squad/current_squad = null
 	///The beacon we will send the supplies
-	var/obj/item/squad_beacon/current_beacon = null
+	var/datum/component/supply_beacon/current_beacon = null
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_console.dm:12]: Absolutely not,do not hold references to components

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 39]
```diff
@@ -9,12 +9,15 @@
 	var/datum/squad/squad = null
 	var/icon_activated = "motion2"
 	var/obj/machinery/camera/beacon_cam = null
+	var/datum/supply_beacon/beacon
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:12]: What is this

File: code/game/objects/machinery/squad_supply/supply_beacon.dm [hunk 40]
```diff
@@ -149,27 +136,76 @@
...
+	UnregisterSignal(parent, COMSIG_UNREGISTER_BEACON)
+	GLOB.supply_beacon_list -= src
+	SEND_SIGNAL(src, COMSIG_BEACON_UNREGISTERED)
+	unregistered_callback?.Invoke()
+	registered = FALSE
+
+/// Remove this datum from the list of datum when the parent is destroyed
+/datum/component/supply_beacon/proc/parent_qdeleting()
+	SIGNAL_HANDLER
+	GLOB.supply_beacon_list -= src
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_beacon.dm:211]: Why not call unregister() here

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Actually why is this even a component

File: code/game/objects/machinery/squad_supply/supply_beacon.dm (hunk 40 above)
Comment (BraveMole) [reply code/game/objects/machinery/squad_supply/supply_beacon.dm:211]: no need, unregister does more thing than this.

File: code/game/objects/machinery/squad_supply/supply_beacon.dm (hunk 36 above)
Comment (BraveMole) [reply code/game/objects/machinery/squad_supply/supply_beacon.dm]: To prevent people from having two antenna helmets in their bag and using them both. That leads to issue cause of similar names

**Comment by DerrickChamberlain** [comment]
I love this, please put this in.

**Comment by DerrickChamberlain** [comment]
Much needed QOL

File: code/game/objects/machinery/squad_supply/supply_console.dm [hunk 41]
```diff
@@ -6,59 +6,68 @@
...
 	var/list/supplies = list()
 	///X offset of the drop, relative to the supply beacon loc
 	var/x_offset = 0
 	///Y offset of the drop, relative to the supply beacon loc
 	var/y_offset = 0
+	/// The pad it is linked to
+	var/obj/structure/supply_drop/drop_pad
 
 /obj/machinery/computer/supplydrop_console/Initialize()
 	. = ..()
```
Comment (TiviPlus) [inline code/game/objects/machinery/squad_supply/supply_console.dm:23]:
```suggestion
	..()
```

File: code/modules/modular_armor/helmets.dm [hunk 42]
```diff
@@ -111,7 +111,30 @@
...
+	var/datum/supply_beacon/beacon
+
+
+/obj/item/helmet_module/antenna/Initialize()
+	. = ..()
+	beacon = new
+
+/obj/item/helmet_module/antenna/Destroy()
+	. = ..()
+	beacon.unregister()
```
Comment (TiviPlus) [inline code/modules/modular_armor/helmets.dm:126]: Why does this not anull it

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Review by TiviPlus — APPROVED** [review]
